### ISSUE TITLE
Rules fail even if there is no change that violates them; unexpected behavior

### ISSUE THREAD

**Issue by cmars** [snyk/sweater-comb#116]
If a rule (typically naming) is violated in both the from and to, Optic CI fails. The problem is, we have some API specs that landed prior to some of our standards being complete, and/or completely implemented in the Spectral rules.

Optic CI can be introduced into such a project if it applies rules to the _changes between --from and --to_; that would have the nice “forcing function” of requiring a team to release a new API version once they want to significantly change a non-compliant one. And, it’d be fairly non-blocking — if they complained, we could always say, “hey, no problem, cut a new version and do what you need to do!“.

To address the special case of "let's run all the rules on everything to see where we're at wrt conformance" we could then tell Optic CI "everything is new". But the normal CI case should be "only apply rules to spec changes in this PR".

To reproduce: See #115. Using fixture introduced there:

`optic-ci compare --from 000-fail-naming.yaml  --to 000-fail-naming.yaml` should PASS and exit 0, because that is effectively "no change". Even though the spec introduced from scratch with `optic-ci compare --to 000-fail-naming.yaml` should FAIL.

**Comment by acunniffe** [comment]
Ok the rules this would happen with are all the `requirement` kinds (mostly naming)

Essentially the functionality you're asking for is a flag that would disable, mask, or change the priority from `must` to `should` for requirement rules that fail in both the `from` spec and the `to` spec. 

- (typical case) If they fail in only `to` spec, they're new issues that must be addressed before commiting
- If they fail in only the `from` spec, they're old issues that are now not showing up in `to`. A fixed requirement failure. It might be fixed in a way that introducing breaks along the way -- so some issues might be traded for another. Ie renaming a field to match the naming rule, is a breaking change is some cases. One issue goes away, another appears. 
- (this new case you're asking for) If they fail in both `from` and `to` they were there before optic was introduced (presumably) or someone said it was ok to skip the requirment, and carrying on into new versions. 


an appropriately named flag for this behavior would be something like
```
--allow-from-spec-requirements-to-fail
```

I wouldn't recommend leaving something like this on all the time though. Maybe just until all the current spec versions are bumped to your new standards. 

This is generally a good story to start thinking about rule lifecycle more broadly.

Does something like this make sense?

**Comment by cmars** [comment]
I don't consider this a new case I'm asking for, and I'm not asking for a new flag.

I consider this to be the core functionality and key differentiator of Optic over other solutions like Spectral, and the expected behavior.

Requirement rules should apply to the contents of the delta. If that delta includes an addition that violates the rule, then CI should fail. If the delta removes it or there is no change, then it shouldn't fail.

This way we can decide how we want to evaluate rules by decided the `from` and `to`. If we want to vet "what's changed in this PR" to gate landing in CI, we compare the target branch `from` to the proposed branch `to`. If we want to see what the state of a given version is, w.r.t compliance, we compare it as if it is all `to` with no `from`. Make sense?

**Comment by acunniffe** [comment]
Thanks for further clarification - I understand why this matters a lot more now and how you’re thinking about the other case (when you want to see where you’re at). 

I have the code working to make this the default behavior. I’m currently testing it to understand the impacts more. 

One thing I’ve already realized:
- if you ever approve / merge past any of these requirement failures they’ll never show up in compare again. Is that what we want? Should they become warnings instead of errors? My worry is that they are silently buried….
- might make sense to have an alternative command to compare that just runs one spec against an empty one (instead of using empty from). It’s more of an ergonomic thing but might make it more clear to users how it works. 
- I can see some cases where running all requirements is what you want to do — maybe a flag to do that would help others users (not important now) 

Tl;dr — this doesn't bring huge changes but I’m happy that we found it now. This being default behavior should make Optic-CI easier to adopt. And since changing these issues is breaking anyhow, it’s not like we’re actually asking users to do something reasonable by showing these errors.

**Comment by acunniffe** [comment]
Ok @cmars I have done a lot of experimentation today and have 2 draft PRs, representing different approaches, to consider.

First let's define the rule selectors and their current expected behaviors:
- `added` -- will apply a rule to an entity that was added to a spec
- `removed` -- will apply a rule to an entity that was removed from the spec
- `changed` -- will apply a rule to an entity that changed (ie field just became required) 
- `requirement` -- a rule that must always hold in the `to` specification (most like spectral rules, select something, apply rule) 

---

- `added` `removed` `changed` -- selectors are natively forward-only (they'll only apply their rule if the author updated that part of the spec. 
- `requirement` -- should really only be used for something that just absolutely must be there. 

---

Essentially the forwards-only version of a `requirement`, what you're asking for above, is a `requiredWhenChanged`. 

I've built this here and made it the default behavior: https://github.com/snyk/sweater-comb/pull/118

After playing with this for an hour I've noticed a few **consequences**
- Rules that check for the presence of something, like the `version` parameter will not work. That's because if something is missing in `from`, it's absence in `to` won't fail. That's a consequence of triggering these rules only on `added` or `changed`. Absence won't trigger the requirement because it wasn't changed by the user. All rules of this kind stop being enforceable. *I know some ways around this....but it adds a fair amount of surface area / changes how rules were written*

> the way around this problem is to run the rule on the operation (or parent), not the header/parameter/field. That'd enforce the rule when the operation was added. Doable, but not necessary in the other approach
 
- If a user merges past a blocked CI issue, it will never show up again and be buried until you run the check again 

--- 

**Alternatively, let's just use the selectors properly** 

We've already built this functionality. `added` `changed` and `removed` rules are meant to do this OOTB. For whatever reason Stephen and I did not use them properly (heck this is all new stuff). 

Here's how we wrote the rule for enforcing property names
```
bodyProperties.requirement.must("have snake case keys", ({ key }) => {
      // TODO: did not find a doc link for this
      const snakeCase = /^[a-z]+(?:_[a-z]+)*$/g;
      expect(snakeCase.test(key)).to.be.ok;
    });
```

This essentially says, "all properties must have a snake case name"
- would trigger on ANY property that's in the `to` spec -- even if it was not changed
- would trigger on ANY property added in a new resource version

Were we to change it to `added`
```
bodyProperties.added.must("have snake case keys", ({ key }) => {
```

Then it would say, "all new properties must have a snake case name"
- would trigger for all properties defined in a new resource version
- would trigger when new properties are added to an existing resource version
- would not trigger on properties in an existing resource version that were named w/o snake case

I made another PR where I went through and tried to tune the rules better: https://github.com/snyk/sweater-comb/pull/119



--- 

**Conclusion & my thoughts** 
- Our existing rules engine will support this use case, if we update some of our `requirement` -> `added` 
- I now agree with @cmars that this should not be a flag.
- I think there's a case for another selector with distinct behavior from our current `requirement`. It would be called `requiredOnChange` and essentially behave as-if you had written the same rule twice with added/change selectors. 
- The original / current behavior of `requirement` is essential if we want to enforce the presence of things like headers/query params/properties. Because something missing in `from` and `to` would never trigger a rule based on change. 

tl;dr -- instead of adding flags or changing core behaviors. I think we just have to update our rules to use the correct selectors. A new selector `requiredOnChange` w/ Casey's suggested behavior makes a lot of sense for certain rules, but if that was the only form of `requirement` it'd make it hard to write certain kinds of rules. 

**Next step** -- sync up. If we go my suggested route, we can walk through the rules together one/by/one (probably in 30) mins and finishing tuning them based on the Snyk use cases. 

I feel like I know 5x more about this whole problem space working through this the past...wow 5 hrs :) Let's chat. Good options all around. We'll get this working Monday.
